muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/ChatController.java

```diff
@@ -1,131 +1,801 @@
+      }
+    }
+    
+    // Notify everyone of a new user
+    
+    String name = userEntityController.getName(userEntity.defaultSchoolDataIdentifier(), true).getDisplayNameWithLine();
+    ChatUserRestModel userRestModel = new ChatUserRestModel(
+        userEntity.getId(),
+        nick,
+        null,
+        getUserType(userEntity),
+        visibility,
+        userProfilePictureController.hasProfilePicture(userEntity),
+        ChatUserPresence.ONLINE);
+    try {
+      if (isStaffMember(userEntity)) {
+
+        // If the new user is staff, everyone may know their real name
+
+        userRestModel.setName(name);
+        Set<Long> userEntityIds = visibility == ChatUserVisibility.STAFF ? listOnlineUserEntityIds(ChatUserType.STAFF) : listOnlineUserEntityIds();
+        userEntityIds.remove(userEntity.getId()); // don't send message to ourselves
+        webSocketMessenger.sendMessage("chat:user-joined", mapper.writeValueAsString(userRestModel), userEntityIds);
+      }
+      else {
+
+        // If the new user is student, only staff may know their real name
+
+        if (visibility == ChatUserVisibility.ALL) {
+          Set<Long> userEntityIds = listOnlineUserEntityIds(ChatUserType.STUDENT);
+          userEntityIds.remove(userEntity.getId()); // don't send message to ourselves
+          if (!userEntityIds.isEmpty()) {
+            webSocketMessenger.sendMessage("chat:user-joined", mapper.writeValueAsString(userRestModel), userEntityIds);
+          }
+        }
+        Set<Long> userEntityIds = listOnlineUserEntityIds(ChatUserType.STAFF);
+        userEntityIds.remove(userEntity.getId()); // don't send message to ourselves
+        if (!userEntityIds.isEmpty()) {
+          userRestModel.setName(name);
+          webSocketMessenger.sendMessage("chat:user-joined", mapper.writeValueAsString(userRestModel), userEntityIds);
+        }
+        
+        // If the new user is student with STAFF visibility, send a nick change to other students (just in case of private convos)
+        
+        if (userRestModel.getType() == ChatUserType.STUDENT && visibility == ChatUserVisibility.STAFF) {
+          userEntityIds = listOnlineUserEntityIds(ChatUserType.STUDENT);
+          userEntityIds.remove(userEntity.getId()); // don't send message to ourselves
+          NickChangeRestModel nickChange = new NickChangeRestModel(userEntity.getId(), nick); 
+          webSocketMessenger.sendMessage("chat:nick-change", mapper.writeValueAsString(nickChange), userEntityIds);
+        }
+      }
+    }
+    catch (JsonProcessingException e) {
+      logger.severe(String.format("Message parse failure: %s", e.getMessage()));
+    }
+  }
+
+  private void handleUserLeave(Long userEntityId, ChatLeaveType type) {
+    
+    // Remove all session data related to the user
+    
+    Set<String> sessionIds = userSessions.get(userEntityId);
+    if (sessionIds != null) {
+      for (String sessionId : sessionIds) {
```

Comment:
Could this have concurrency problems, the sessionIds set could be modified/iterated by two threads at the same time? Not a likely occurence for sure.